backend/arm64/isa/arm64_ast.ml

```diff
@@ -0,0 +1,597 @@
+      | AL -> "al"
+  end
+
+  module Rounding_mode = struct
+    type t =
+      | M
+      | P
+      | Z
+      | X
+      | N
+
+    let to_string t =
+      match t with M -> "m" | P -> "p" | Z -> "z" | X -> "x" | N -> "n"
+  end
+
+  (* Don't split the type into base and neon to make emit easier to read and
+     avoid the need to copy all the instructions in the DSL. *)
+  type t =
+    (* base *)
+    | NOP
+    | ADD
+    | SUB
+    | MUL
+    | DIV
+    | AND
+    | OR
+    | XOR
+    | LSL
+    | LSR
+    | ASR
+    | CLZ
+    | RBIT
+    | CNT
+    | SMULH
+    | UMULH
+    | B of Cond.t
+    | BL
+    | BLR
+    | CBNZ
+    | CBZ
+    | CSEL
+    | CSET
+    | SXTB
+    | SXTH
+    | SXTW
+    | UXTB
+    | UXTH
+    (* neon *)
+    | MOV
+    | MOVI
+    | FMOV
+    | FADD
+    | FSUB
+    | FMUL
+    | FDIV
+    | FNMUL
+    | FMADD
+    | FNMADD
+    | FMSUB
+    | FNMSUB
+    | FNEG
+    | FABS
+    | FSQRT
+    | FCVT
+    | FCVTZS
+    | FCVTNS
+    | SCVTF
+    | FRINT of Rounding_mode.t
+    | FRINT64 of Rounding_mode.t
+    | FMIN
+    | FMAX
+    | ZIP1
+    | ZIP2
+    | FCMP
+    | FCSEL
+
+  (* CR gyorsh: can some of this be automatically generated from the type? *)
+  let to_string t =
+    match t with
+    | NOP -> "nop"
+    | ADD -> "add"
+    | SUB -> "sub"
+    | MUL -> "mul"
+    | DIV -> "div"
+    | AND -> "and"
+    | OR -> "or"
+    | XOR -> "xor"
+    | LSL -> "lsl"
+    | LSR -> "lsr"
+    | ASR -> "asr"
+    | CLZ -> "clz"
+    | RBIT -> "rbit"
+    | CNT -> "cnt"
+    | SMULH -> "smulh"
+    | UMULH -> "umulh"
+    | B c -> "b." ^ Cond.to_string c
+    | BL -> "bl"
+    | BLR -> "blr"
+    | CBNZ -> "cbnz"
+    | CBZ -> "cbz"
+    | CSEL -> "csel"
+    | CSET -> "cset"
+    | SXTB -> "sxtb"
+    | SXTH -> "sxth"
+    | SXTW -> "sxtw"
+    | UXTB -> "uxtb"
+    | UXTH -> "uxth"
+    (* neon *)
+    | MOV -> "mov"
+    | MOVI -> "movi"
+    | FMOV -> "fmov"
+    | FADD -> "fadd"
+    | FSUB -> "fsub"
+    | FMUL -> "fmul"
+    | FDIV -> "fdiv"
+    | FNMUL -> "fnmul"
+    | FMADD -> "fmadd"
+    | FNMADD -> "fnmadd"
+    | FMSUB -> "fmsub"
+    | FNMSUB -> "fnmsub"
+    | FNEG -> "fneg"
+    | FABS -> "fabs"
+    | FSQRT -> "fsqrt"
+    | FCVT -> "fcvt"
+    | FCVTZS -> "fcvtzs"
+    | FCVTNS -> "fcvtns"
+    | SCVTF -> "scvtf"
+    | FRINT rm -> "frint" ^ Rounding_mode.to_string rm
+    | FRINT64 rm -> "frint64" ^ Rounding_mode.to_string rm
+    | FMIN -> "fmin"
+    | FMAX -> "fmax"
+    | ZIP1 -> "zip1"
+    | ZIP2 -> "zip2"
+    | FCMP -> "fcmp"
+    | FCSEL -> "fcsel"
+
+  let print ppf t = Format.fprintf ppf "%s" (to_string t)
+end
+
+module Operand = struct
+  module Imm = struct
+    (* int is big enough for all instruction encodings *)
+    type t = int
+
+    let print ppf t = Format.fprintf ppf "#%d" t
+  end
+
+  module Extend = struct
+    module Kind = struct
+      type t =
+        | UXTB (* 000 *)
+        | UXTH (* 001 *)
+        | UXTW (* 010 *)
+        | UXTX
+        (* 011 *)
```

Comment:
```suggestion
        | UXTX (* 011 *)
```

Reply:
hmm, this was supposed to be auto-formatted. I am moving the file to `backend` to address your other comment, it'll help with autoformatting too.